### ISSUE TITLE
Long-term Realtime Database connection with ChildListener

### ISSUE THREAD

**Issue by mrnrone** [firebase/firebase-cpp-sdk#38]
### Please fill in the following fields:
Pre-built SDK from the [website](https://firebase.google.com/download/cpp)
Firebase C++ SDK version: 6.8.0
Firebase plugins in use (Auth, Database, etc.): App, Auth, Realtime Database
Additional SDKs you are using (Facebook, AdMob, etc.):  Qt 5.12.5
Platform you are using the C++ SDK on (Mac, Windows, or Linux): Windows 10
Platform you are targeting (iOS, Android, and/or desktop): Desktop

### Please describe the issue here:
1. Sign-in existing user with Auth (email + password)
2. Get database reference for a selected root node
3. Add ChildListener to this node
4. Desktop application is running: ChildListener is triggered for every child nodes
5. Add some child node from a mobile application: ChildListener is triggered for every new child nodes
6. Both mobile and desktop applications are running for more than 1 hour in "idle state". (always connected, no sleep mode, no new child nodes are being sent from mobile app)
7. Add some new child node from a mobile application:  these nodes can be shown in Firebase web console but ChildListener is NOT triggered anymore on desktop! Re-login and new database reference is required in order to get them. 

Q: What is the best practice to always get the new child nodes on desktop? How long is the connection alive?

### Please answer the following, if applicable:
Have you been able to reproduce this issue with just the [Firebase C++ quickstarts](https://github.com/firebase/quickstart-cpp) ?
Did not tried.

What's the issue repro rate? (eg 100%, 1/5 etc)
-

**Comment by alexames** [comment]
Hi @mrnrone, 

We would expect that the connection should stay alive as long as the app is open. We have a few theories about what might be causing it to disconnect, but we'd like some more data. Would you mind turning on verbose logging, reproducing the issue, and then uploading the log? That will let us see what is happening when it fails.

**Comment by mrnrone** [comment]
Hi Alex,

please let me know how I can turn on verbose logging and where the log file
I can found.

Thanks

**Comment by alexames** [comment]
You will want to use two functions to turn on logging, one for [firebase in general](https://firebase.google.com/docs/reference/cpp/namespace/firebase#setloglevel), and one for [database specifically](https://firebase.google.com/docs/reference/cpp/class/firebase/database/database#set_log_level):

```
::firebase::SetLogLevel(::firebase::kLogLevelVerbose);
my_database->set_log_level(::firebase::kLogLevelVerbose);
```

Logs will be output to standard out, so you can just capture them and save them to a file.

**Comment by mrnrone** [comment]
This part of the log might be useful:

... 
DEBUG: ID token changed. Notifying 1 listeners.
DEBUG: [conn_1] websocket message received
DEBUG: [conn_1] websocket message received
DEBUG: [conn_1] ProcessMessage (length: 79)
DEBUG: [conn_1] received data message
DEBUG: [pc_0] handleServerMessage ac {"d":"Auth token is expired.","s":"expired_token"}
DEBUG: [pc_0] Auth token revoked: 2 (Auth token is expired.)
DEBUG: Starting transaction.
DEBUG: Transaction completed.
DEBUG: [conn_1] Closing connection. Reason: 0
DEBUG: [pc_0] Got on disconnect due to 0

End of log

Please let me know what listener(s) should be used to get notification about the expired token? Should I simply re-login with authentication module and reconnect to the database in this case?

**Comment by mrnrone** [comment]
I tried to update auth token in different ways like:

Call periodically currentUser->GetToken(true) method
or
Call periodically currentUser->Reauthenticate(userCredential) method 

but in both cases the auth token expired after 1 hour.  

As a workaround in every 30 minutes I sing out, disconnect from database and do a re-sign in and re-connection to the database and skip previous messages. It's not the best solution, but it works.

**Comment by DellaBitta** [comment]
We believe that this is an issue with the auth token not propagating to the database service.  Thanks for the report!

**Comment by FrankNine** [comment]
Hello, we are experiencing similar issue while using Firebase Unity plugin 6.8.0 which I believe is a wrapper of Firebase C++ plugin 6.8.0. 

We built a chat system on Realtime Database and manage Presence according to Firebase Realtime Database [samples](https://firebase.google.com/docs/database/web/offline-capabilities#section-presence). We add Presence when user is connected to the Realtime Database and remove Presence with OnDisconnect. But if the user move our app to background, then after an hour the user would lost his/her authentication and got access denied when executing OnDisconnect (auth.uid not equal uid)
```
{
  "rules" : {
    "presence" : {
      "$uid" : {
        ".write" : "$uid === auth.uid"
      }
    }
  }
}
```

I would like to ask is there an update about this issue? Thank you.

Frank

**Comment by chkuang-g** [comment]
Hi folks,

We have patched for the authentication part of this issue and it should be included in the next release.
With the fix, if the client idle for an hour, it would reconnect with the new auth token and re-establish all listener event.
The catch is: it would still trigger OnDisconnect event.  We are working on the part to send the fresh auth token to database server periodically in order to prevent the disconnection.

@FrankNine Are you experiencing "access denied when executing OnDisconnect" issue on desktop or on Android/iOS devices?  If on mobile devices, that would sound like an issue on mobile SDK.  Could you confirm this?  Thank you.

Shawn

**Comment by FrankNine** [comment]
We are experiencing it on Android devices. I was trying to capture full log but it seems log change a bit after 6.10.0 that I couldn't find that OnDisconnect error, but the presence duplication due to OnDisconnect has not authentication to write still happened.

**Comment by chkuang-g** [comment]
@FrankNine 

Thank you for reporting this issue!  However, yours seems to be different from what this thread is for, which is a bug on desktop or in editor.

To help us pinpoint and manage the issue easier, could you kindly open a new issue with detail about your problem? If possible, could you try to reproduce the behavior using our [quick-start project](https://github.com/firebase/quickstart-unity/tree/master/database/testapp)?

Thank you,
Shawn

**Comment by patm1987** [comment]
The original issue should be fixed as of [6.11.0](https://firebase.google.com/support/release-notes/cpp-relnotes#version_6110_-_february_20_2020). Let us know if it still occurs!

--Patrick
